payment/watcher.py

```diff
@@ -1,80 +1,158 @@
+        return seq
 
+    def on_payment(self, address, order_id, tx: SimplifiedTransaction):
+        """handle a new payment from/to an address."""
+        payment = Payment.from_blockchain(tx)
+        log.info(f'got payment for order {order_id} for address {address}', payment=payment)
+        statsd.inc_count('payment_observed',
+                         payment.amount,
+                         tags=['app_id:%s' % payment.app_id,
+                               'address:%s' % address])
 
-def get_last_cursor():
-    cursor = CursorManager.get()
-    if not cursor:
-        @retry(5, 0.2)
-        def get_from_horizon():
-            return Blockchain.get_last_cursor()
-        cursor = get_from_horizon()
-        log.info('got payment cursor from horizon', cursor=cursor)
-        CursorManager.save(cursor)
+        self.enqueuer.enqueue_payment_callback(self.webhook, address, payment)
+        self.remove(address, order_id)
 
-    return cursor
+    async def worker(self):
+        """Poll blockchain and apply callback on watched address. run until stopped."""
+        ledger_seq = await self.get_last_ledger_sequence()
+        retry_count = 0
+        while True:
+            if retry_count == 0:  # Otherwise, stay with the old start_time
+                start_t = time.time()
+            try:
+                addresses = self.watched_addresses.keys()
+                log.debug(f'Watching for ledger {ledger_seq}, try: #{retry_count + 1}')
 
+                try:
+                    txs = await self.bc_manager.get_txs_per_ledger(ledger_seq)
+                    retry_count = 0
+                except ResourceNotFoundError:
+                    log.debug(f'Ledger {ledger_seq} not found')
+                    retry_count += 1
+                    if retry_count >= RETRIES_BEFORE_EXCEPTION:
+                        raise
 
-def on_payment(address, payment):
-    """handle a new payment from an address."""
-    log.info('got payment', address=address, payment=payment)
-    statsd.inc_count('payment_observed',
-                     payment.amount,
-                     tags=['app_id:%s' % payment.app_id,
-                           'address:%s' % address])
+                    log.debug(f'retrying in {WAIT_WHEN_NOT_FOUND} seconds')
+                    await asyncio.sleep(WAIT_WHEN_NOT_FOUND)
+                    continue
 
-    for service_id in Watcher.get_subscribed(address):
-        enqueue_payment_callback(Service.get(service_id), address, payment)
+                if len(txs) != 0:
+                    """
+                    get_txs_per_ledger returns a list of responses from horizon.
+                    Building the Transaction object, which includes decoding the envelops of potentially hundreds of txs 
+                    is cpu intensive and can take a lot of time, (~80ms per 200 txs with 1 op each)
+                    we shouln't block the event loop for that long, so run it in a separate process
```

Comment:
is it process or thread? and if process than why an entire process is needed?

Reply:
Process, usually you use processes for CPU intensive load, and threads for IO, also, with processes you can get actual parallelism, threads cant give you that in python because of the GIL